File: vjezbe/vjezba_3/zd_6.cpp
// CSES Chessboard and Queens

#include <iostream>
#include <vector>
#include <string>

using namespace std;

int countWays = 0;
vector<string> board(8);
vector<bool> col(8, false), diag1(15, false), diag2(15, false);

void solve(int row) {
    if (row == 8) {
        countWays++;
        return;
    }
    for (int c = 0; c < 8; c++) {
        if (board[row][c] == '.' && !col[c] && !diag1[row + c] && !diag2[row - c + 7]) {
            col[c] = diag1[row + c] = diag2[row - c + 7] = true;
            solve(row + 1);
            col[c] = diag1[row + c] = diag2[row - c + 7] = false;
        }
    }
}

int main() {
    for (int i = 0; i < 8; i++) {
        cin >> board[i];
    }
    solve(0);
    cout << countWays << endl;
    return 0;
}

File: vjezbe/vjezba_3/zd_6.py
def count_ways(board):
    cols = [False] * 8        
    diag1 = [False] * 15      
    diag2 = [False] * 15      

    def place_queen(row):
        if row == 8:  
            return 1
        ways = 0
        for col in range(8):
            if board[row][col] == '.' and not cols[col] and not diag1[row - col + 7] and not diag2[row + col]:
                cols[col] = diag1[row - col + 7] = diag2[row + col] = True
                ways += place_queen(row + 1)
                cols[col] = diag1[row - col + 7] = diag2[row + col] = False
        return ways

    return place_queen(0)

board = [input().strip() for _ in range(8)]
print(count_ways(board))
